### PR TITLE
Feature/gh 9785 create config

### LINKED ISSUE #9785: Service for calling CPQ create configuration, interceptor and access storage service

### PR DESCRIPTION
**Pull request by ChristophHi**

### CODE REVIEW

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    serviceUnderTest = TestBed.inject(
+      CpqConfiguratorRestService as Type<CpqConfiguratorRestService>
+    );
+  });
+
+  afterEach(() => {
+    httpMock.verify();
+  });
+
+  it('should fetch a token and use it to init a configuration', () => {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.spec.ts:59]: This test is failing

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { CPQ_CONFIGURATOR_NORMALIZER } from './cpq-configurator.converters';
+
+@Injectable({ providedIn: 'root' })
+export class CpqConfiguratorRestService {
+  constructor(
+    protected http: HttpClient,
+    protected converterService: ConverterService
+  ) {}
+
+  createConfiguration(
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.ts:22]: Here I would suggest a method comment, for public we are generally encouraged to do so (didn't do it always), but here it specifically makes sense as the method does 2 http calls

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+              ...resultConfiguration,
+              configId: configCreatedResponse.configurationId,
+            };
+          })
+        );
+      })
+    );
+  }
+
+  protected callConfigurationsInit(
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.ts]: 'configurationInit'?

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  ): Observable<Cpq.ConfigurationCreatedResponseData> {
+    return this.http.post<Cpq.ConfigurationCreatedResponseData>(
+      `${CPQ_CONFIGURATOR_VIRTUAL_ENDPOINT}/api/configuration/v1/configurations`,
+      {
+        ProductSystemId: productSystemId,
+      }
+    );
+  }
+
+  protected callConfigurationsDisplay(
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.service.ts]: 'configurationDisplay'?

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import { expand, filter, shareReplay, switchMap } from 'rxjs/operators';
+import { CpqAccessData } from './cpq-access-data.models';
+import { CpqAccessLoaderService } from './cpq-access-loader.service';
+
+const ONE_DAY = 24 * 60 * 60 * 1000;
+
+@Injectable({ providedIn: 'root' })
+export class CpqAccessStorageService {
+  /* We should stop using/sending a token shortly before expiration,
+   * to avoid that it is actaully expired when evaluated in the target system */
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts]: typo

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+  constructor(
+    protected cpqAccessLoaderService: CpqAccessLoaderService,
+    @Inject(CpqAccessStorageService.TOKEN_EXPIRATION_BUFFER)
+    protected EXPIRATION_BUFFER: number
+  ) {}
+
+  protected cpqAccessData: Observable<CpqAccessData>;
+
+  private static defaultValue() {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts]: Please never use 'private', and have this as default configuration (not publicly exposed. See e.g. /feature-libs/product/configurators/rulebased/components/config/default-message-config.ts.
Customers then can easily override

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import { Inject, Injectable, InjectionToken } from '@angular/core';
+import { Observable, timer } from 'rxjs';
+import { expand, filter, shareReplay, switchMap } from 'rxjs/operators';
+import { CpqAccessData } from './cpq-access-data.models';
+import { CpqAccessLoaderService } from './cpq-access-loader.service';
+
+const ONE_DAY = 24 * 60 * 60 * 1000;
+
+@Injectable({ providedIn: 'root' })
+export class CpqAccessStorageService {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts:41]: How can we trigger a new re-initialization of the token, e.g. if logout/login occured?
(we discussed this several times, one way would be to manage the token via a BehaviourSubject, there might be others)

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-configurator-rest.interceptor.spec.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  CpqConfiguratorRestInterceptor,
+  CPQ_CONFIGURATOR_VIRTUAL_ENDPOINT,
+} from './cpq-configurator-rest.interceptor';
+
+describe('CpqConfiguratorRestInterceptor', () => {
+  let interceptorUnderTest: CpqConfiguratorRestInterceptor;
+
+  const nonCPQRequest: HttpRequest<any> = new HttpRequest(
+    'GET',
+    'https://www.exmaple.com'
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-configurator-rest.interceptor.spec.ts]: typo

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-configurator-rest.interceptor.spec.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      .intercept(cpqRequest, mockedNextHandler)
+      .subscribe(() => {
+        expect(capturedRequest.url).toBe(
+          cpqAccessData.endpoint + '/api/whatever'
+        );
+        done();
+      });
+  });
+
+  it('should add authorization header to cpq related requetes', (done) => {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-configurator-rest.interceptor.spec.ts]: typo

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-configurator-rest.interceptor.spec.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+      .intercept(cpqRequest, mockedNextHandler)
+      .subscribe(() => {
+        expect(capturedRequest.headers.get('Authorization')).toBe(
+          'Bearer TOKEN'
+        );
+        done();
+      });
+  });
+
+  it('should add cookieless header to cpq related requetes', (done) => {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-configurator-rest.interceptor.spec.ts]: typo

File: feature-libs/product/configurators/rulebased/root/interceptor/rulebased-configurator-interceptor.module.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  imports: [CommonModule],
+  providers: [
+    {
+      provide: HTTP_INTERCEPTORS,
+      useClass: CpqConfiguratorRestInterceptor,
+      multi: true,
+    },
+  ],
+})
+export class RulebasedConfiguratorInterceptorModule {}
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/rulebased-configurator-interceptor.module.ts]: Better rename to CpqConfiguratorInterceptorModule

File: feature-libs/product/configurators/rulebased/rulebased-configurator.module.ts [hunk 12]
```diff
@@ -14,4 +14,10 @@ import { VariantConfiguratorOccModule } from './occ/variant/variant-configurator
     CpqConfiguratorRestModule,
   ],
 })
-export class RulebasedConfiguratorModule {}
+export class RulebasedConfiguratorModule {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/rulebased-configurator.module.ts]: Why this change?

Comment (Uli-Tiger) [reply feature-libs/product/configurators/rulebased/rulebased-configurator.module.ts]: seems to be leftover from my various tries to get the interceptor working in a lazy environment - i'll revert.

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts (hunk 7 above)
Comment (Uli-Tiger) [reply feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts:41]: @ChristophHi Token fetching is  now started/stopped depending on login state.
Logging out will now unsubsribe from the token-access service, while logging in will subscribe to it.
subscribing to the login state is triggered by the first attempt to fetch a token, so this happens only if cpqConfigurator is used.

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts [hunk 13]
```diff
@@ -0,0 +1,195 @@
+import { Injectable } from '@angular/core';
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts:1]: Test coverage is too low, see coverage report

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.adapter.ts [hunk 14]
```diff
@@ -1,25 +1,18 @@
...
-import { map } from 'rxjs/operators';
+import { Observable, of } from 'rxjs';
+import { delay, map } from 'rxjs/operators';
 import { RulebasedConfiguratorAdapter } from '../core/connectors/rulebased-configurator.adapter';
 import { Configurator } from '../core/model/configurator.model';
-import { CpqAccessLoaderService } from '../occ/cpq/cpq-access-loader.service';
+import { CpqConfiguratorRestService } from './cpq-configurator-rest.service';
 
 @Injectable()
 export class CpqConfiguratorRestAdapter
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.adapter.ts:11]: Test coverage is way too low (64 % statement coverag)

File: feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  protected initCpqAccessObservable() {
+    this.cpqAccessDataCache = new BehaviorSubject(EXPIRED_TOKEN);
+    this.cpqAccessObservable = this.cpqAccessDataCache.pipe(
+      // Never expose expired tokens - either cache was invalidated with expired token,
+      // or the cached one expired before a new one was fetched.
+      filter((data) => !this.isTokenExpired(data))
+    );
+    this.authService.isUserLoggedIn().subscribe((loggedIn) => {
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/root/interceptor/cpq-access-storage.service.ts]: Even if technically not required: 
Let's add pipe(dinstinctUntilChanged), because authService.isUserLoggedIn() might change its behaviour, and in addition everyone, when viewing this code, needs to check that it does distinctUntilChanged internally..

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+import { Converter } from '@spartacus/core';
+import { Configurator } from './../core/model/configurator.model';
+import { Cpq } from './cpq.models';
+
+@Injectable({ providedIn: 'root' })
```
Comment (ChristophHi) [inline feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts]: Must not be provided in root! Is explicitly provided in rest module

**Review by ChristophHi — COMMENTED** [review]
I have 2 minor comments, additionally the test coverage needs to be enhanced

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.adapter.ts (hunk 14 above)
Comment (Uli-Tiger) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-rest.adapter.ts:11]: deleted some unused dummy coding an added an additional test. should be 100% now

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts (hunk 13 above)
Comment (AlexanderKuschner) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts:1]: done

File: feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts (hunk 16 above)
Comment (AlexanderKuschner) [reply feature-libs/product/configurators/rulebased/cpq/cpq-configurator-normalizer.ts]: done
